src/lib/Syntax.hs

```diff
@@ -153,23 +164,30 @@ scalarTableBaseType t = case t of
 
 
 type Label = String
-newtype LabeledItems a = LabeledItems (M.Map Label [a])
+newtype LabeledItems a = LabeledItems (M.Map Label (NE.NonEmpty a))
```

Comment:
Can you please add a little comment that describes what this ADT represents? Do I understand correctly, that the idea here is that we keep all the fields of a given label in order, to make sure we keep the scopes right? By the way, would it make sense to put in some extra care to restore the original field order when printing (doesn't have to happen in this PR)?

Reply:
Sure, I can add a comment.

Regarding "original field order", maybe that's something worth having a bigger discussion about. Right now field order is thrown away during parsing; fields are always ordered in alphabetical order after that. This is mainly so that `{a:A, b:B}` and `{b:B, a:A}` unify with each other, and so that you can extract either `a` or `b` in either order. This ADT represents the "canonical form" of the fields, such that we have an unordered collection of labels (treated as an alphabetical order during iteration) and an ordered collection of values for each label (for repeated fields, which DO have a definite ordering).

An alternative approach, I guess, would be to actually treat the order of the fields as part of the type? But then two records with different orders would be different types, and we'd have to cast between them, either explicitly in the concrete syntax or implicitly based on some set of rules. I don't know how that interacts with extensibility; I think most extensible record designs don't treat order as important.

I guess there's a third, purely cosmetic approach, where we associate the labeled items with an optional field order that is ignored by the typechecker? When unifying two things we would just arbitrarily pick one of the orders if they conflict. This would then be used to print things, but not affect the runtime representation? Although that might be a bit confusing because the runtime representation is what determines the iteration order of the record when used as an index set...

Either way probably any change like this should go in a different PR, I think.

Reply:
Yeah I definitely didn't mean to suggest that order of fields should matter, only that we might do a best effort job to preserve it for printing. When it comes to PRs, many of the things that I pointed out could be separated so don't feel super pressured to fix it right away, but this is a convenient place to discuss them.